### PR TITLE
Models api

### PR DESCRIPTION
**Pull request by Lucaskabela**
Make a lot of changes to the Models API and network code.  At a high level, this includes:

- Refactoring AutoencoderNetwork to only have the method stubs
- Making TextRNN, SpeechRNN, TextTransformer, and SpeechTransformer the only classes implementing this
- Adding a Linear to SpeechPostnet (for EOS prediction)
- Adding <SOS> symbol for text
- Training code for speech and text autoencoders
- Concrete (and working) implementations for the TextRNN and SpeechRNN
- Move notebooks to their own folder
- Add saving and loading functionality

### CODE REVIEW

File: src/data/symbols.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 _pad        = '_'
 _eos        = '~'
+_sos        = '-'
```
Comment (liaojh1998) [inline src/data/symbols.py:14]: Just making sure, this is possible because there's no '-' in phonemes right?

File: src/module.py [hunk 2]
```diff
@@ -141,18 +142,21 @@ def __init__(self, num_mels, num_hidden, p=0.1):
...
+        self.stop_linear = nn.Linear(num_mels, 1)
 
     def forward(self, input_):
         """
         :param input_: Tensor of mel-spectrogram output from decoder,
                        these should be of dimensions [batch, num_mels, length]
         """
         # Causal Convolution (for auto-regressive)
+        stop_pred = self.stop_linear(input_)
+        input_ = input_.permute(0, 2, 1)
```
Comment (liaojh1998) [inline src/module.py]: Let's change the comment about dimensions to `[batch, length, num_mels]` for this.

File: src/network.py [hunk 3]
```diff
@@ -120,25 +101,225 @@ def forward(self, enc_output):
         temp2 = self.dropout(self.non_linear(self.fc2(temp)))
         return self.fc3(temp2)
     
-class SpeechTransformer(SpeechNet):
+class SpeechTransformer(AutoEncoderNet):
     # TODO: Fill in with pre/post needed and enc/dec
-    def __init_(self):
+    def __init__(self, args):
         super(SpeechTransformer, self).__init__()
```
Comment (liaojh1998) [inline src/network.py:107]: Nice, I'll add to this once the PR is merged.

File: src/train.json [hunk 4]
```diff
@@ -1,15 +1,25 @@
...
+    "e_p": 0.5,
+    "e_layers": 1,
+    "e_bi": false,
+    "d_in": 256,
+    "d_out": 128,
+    "d_p": 0.3,
+    "d_layers": 1,
+    "d_attn": false,
+    "grad_clip": 1.0
+    "train_discriminator": false
```
Comment (liaojh1998) [inline src/train.json]: Missing a comma, and let's add comma to the end of every line.
```
    "grad_clip": 1.0,
    "train_discriminator": false,
```

File: src/network.py [hunk 5]
```diff
@@ -120,25 +101,225 @@ def forward(self, enc_output):
...
+        self.postnet = SpeechPostnet(args.num_mels, args.hidden)
+    
+
+    def preprocess(self, raw_input_tensor):
+        # raw_input_tensor should be a LongTensor of size [batch_size x seq_len x 1]
+        # should already be padded as well
+        # Get a mask of 0 for no padding, 1 for padding of size [batch_size x seq_len]
+        # NOTE: I do not think this is correct for mel... but check if all are padding in num_mels
+        pad_mask = torch.all(raw_input_tensor.eq(PAD_IDX), dim=-1).float().squeeze() 
+        return self.prenet(raw_input_tensor), pad_mask
```
Comment (liaojh1998) [inline src/network.py:128]: I thought the padding came at the end of every sequence to extend sequence length and match them into 1 batch. I don't think there's 0 for anything else but we should be wary.

File: src/network.py [hunk 6]
```diff
@@ -120,25 +101,225 @@ def forward(self, enc_output):
...
+        stop_lens = torch.zeros(batch_size).to(enc_output.device)
+        
+        # get a all 0 frame for first timestep
+        input_ = torch.zeros((batch_size, 1, self.postnet.num_mels), device=enc_output.device)
+        i = 0
+        keep_gen = torch.all(stop_lens.neq(0)) and i < max_len
+
+        while keep_gen:
+            dec_out, stop_pred, hidden_state = self.decode(input_, hidden_state, enc_output, enc_ctxt_mask)
+            stops.apped(stop_pred)
```
Comment (liaojh1998) [inline src/network.py]: `stops.append()`

File: src/network.py [hunk 7]
```diff
@@ -120,25 +101,225 @@ def forward(self, enc_output):
...
-class SpeechTransformer(SpeechNet):
+class SpeechTransformer(AutoEncoderNet):
     # TODO: Fill in with pre/post needed and enc/dec
-    def __init_(self):
+    def __init__(self, args):
         super(SpeechTransformer, self).__init__()
 
 
-class SpeechRNN(SpeechNet):
+class SpeechRNN(AutoEncoderNet):
```
Comment (liaojh1998) [inline src/network.py:110]: Is there a `.forward()` for this class? Would it just be like:
```python3
def forward(self, input_):
    representation, ... = self.encode(input_)
    output, ... = self.decode_sequence(representation)
    return output
```

File: src/network.py [hunk 8]
```diff
@@ -120,25 +101,225 @@ def forward(self, enc_output):
...
-class TextTransformer(TextNet):
+class TextTransformer(AutoEncoderNet):
     # TODO: Fill in with pre/post needed and enc/dec
-    def __init_(self):
+    def __init__(self, args):
         super(TextTransformer, self).__init__()
 
 
-class TextRNN(TextNet):
+class TextRNN(AutoEncoderNet):
```
Comment (liaojh1998) [inline src/network.py:229]: Let's add a `.forward()` here too.

File: src/train.py [hunk 9]
```diff
@@ -23,27 +30,55 @@ def autoencoder_loss(output, target, speech=False):
...
+
+def eval_speech_auto(target, hidden_state, enc_output, enc_ctxt_mask, text_model):
+    res = text_model.infer_sequence(hidden_state, enc_output, enc_ctxt_mask, max_len=target.shape[1])
+    loss = loss_fn(res, target).item()
+    return loss
+
+def evaluate(text_model, speech_model, valid_dataset):
+    """
+        Expect validation set to have paired speech & text!
+        We evaluate on 4 metrics:  autoencoder text loss, autoencdoer speech loss,
```
Comment (liaojh1998) [inline src/train.py]: spelling nit on `autoencdoer`

File: src/train.py [hunk 10]
```diff
@@ -61,6 +96,179 @@ def train(args):
...
+        avg_l = np.mean(losses)
+        print("epoch %-3d \t loss = %0.3f \t" % (epoch, avg_l))
+        # if validation < best:
+        # TODO: Add model (and optimizer) saving for reloading training
+        #     print("Saving model!")
+        #     best = validation
+        #     model.save_model()
+    return model
+
+def train(args):
```
Comment (liaojh1998) [inline src/train.py]: Probably dead code here, this function and its contents can be removed. The real `train()` loop is below this one. From what I'm seeing, you're going to call `train_text_auto()` and `train_speech_auto()` in the train loop, right?

Probably change `train_*_auto()` to include a dataloader parameter to pass the unsupervised dataloader to them.

Are we going to have separate functions for cross-modal training?

**Review by liaojh1998 — COMMENTED** [review]
Looks good. Which functions are you currently testing out?

**Review by liaojh1998 — CHANGES_REQUESTED** [review]
Let's make sure the code for training for the autoencoders are working before we merge this.

File: src/data/symbols.py (hunk 1 above)
Comment (Lucaskabela) [reply src/data/symbols.py:14]: Good question @billyang98

File: src/train.json (hunk 4 above)
Comment (Lucaskabela) [reply src/train.json]: For some reason, the trailing comma for JSON file gave me an error (although I think it is just linting related).  So for compatibility purposes with different environments, I think we should leave it out: https://stackoverflow.com/questions/201782/can-you-use-a-trailing-comma-in-a-json-object#:~:text=No.,you%20shouldn't%20include%20it.

File: src/network.py (hunk 5 above)
Comment (Lucaskabela) [reply src/network.py:128]: This is my understanding as well - this mask will effectively check all entries in the last dimension are PAD, and if that is the case then the time step was a padding timestep.  0's do not get used in anything else for text for sure, but for log_mel they are also used for the 0th timestep (like SOS)

File: src/network.py (hunk 7 above)
Comment (Lucaskabela) [reply src/network.py:110]: Yeah I need to add a forward here.  Consider it done!

File: src/train.py (hunk 10 above)
Comment (Lucaskabela) [reply src/train.py]: The train_text_auto() and train_speech_auto are purely for ensuring the text and speech autoencoder work independently rn. They will be removed once train() is done, and replaced with something like `text_auto_step` which takes a batch and the text model and does one iteration of text_autoencoder training

File: src/data/symbols.py (hunk 1 above)
Comment (billyang98) [reply src/data/symbols.py:14]: Yup

File: src/network.py [hunk 11]
```diff
@@ -120,25 +101,235 @@ def forward(self, enc_output):
...
+        dec_output, dec_hidden = self.decoder(input_, hidden_state, enc_output, enc_ctxt_mask)
+        return self.postprocess(dec_output), dec_hidden
 
+    def postprocess(self, dec_output, distrib=False):
+        return self.postnet(dec_output)
 
-class TextTransformer(TextNet):
+    def forward(self, input_, mel_input):
+        encoder_outputs, latent_hidden, pad_mask = model.encode(input_)
+        pred, stop_pred = model.decode_sequence(mel_input, latent_hidden, encoder_outputs, pad_mask)
```
Comment (liaojh1998) [inline src/network.py]: Should this be `model` instead of `self`?

File: src/network.py [hunk 12]
```diff
@@ -120,25 +101,235 @@ def forward(self, enc_output):
...
+        seq_lens = seq_lens.masked_fill(seq_lens == 0, len(outputs))
+        pad_mask = sent_lens_to_mask(seq_lens, len(outputs))
+
+        res = torch.stack(outputs, dim=1).squeeze(2)
+        res = res.masked_fill(pad_mask, PAD_IDX)
+        return res
+
+    def forward(self, input_):
+        encoder_outputs, latent_hidden, pad_mask = model.encode(input_)
+        pred = model.decode_sequence(input_, latent_hidden, encoder_outputs, pad_mask)
```
Comment (liaojh1998) [inline src/network.py]: Same case here, `model` instead of `self`?

File: src/network.py (hunk 5 above)
Comment (liaojh1998) [reply src/network.py:128]: Interesting, is there a reason why they used it for the 0th timestep? I don't think it's too much of an issue to have a PAD at the start though.

File: src/train.json (hunk 4 above)
Comment (liaojh1998) [reply src/train.json]: Cool, I didn't know that, I guess Python allows it nicely.

Anyhow, don't forget the comma for `"grad_clip": 1.0`.

**Review by liaojh1998 — CHANGES_REQUESTED** [review]
Looking good! I added some questions on syntaxes.

File: src/network.py (hunk 5 above)
Comment (Lucaskabela) [reply src/network.py:128]: I think its because in using the mel filters, 0 represents silence, and any other value would be noise.  Its only used on the 0th timestep for the `mel_input` variable though, which is just the `mel` variable but with this 1 timestep of silence for conditioning on

File: src/network.py (hunk 11 above)
Comment (Lucaskabela) [reply src/network.py]: Yeah it absolutely should be, I just copied and pasted this from the training code lol

**Review by liaojh1998 — APPROVED** [review]
Looks great! Thank you for working on this. I can also get it tested if there's a script to run for training a part of it.

File: src/network.py (hunk 5 above)
Comment (liaojh1998) [reply src/network.py:128]: Okay, that consistency is nice!

**Review by billyang98 — APPROVED** [review]
